0012-BIRS-distributed-scheduler.md

```diff
@@ -0,0 +1,463 @@
+message ListJobsRequest {
+    // The id of the application (app_id) for which to list jobs.
+    string app_id = 1;
+}
+
+// ListJobsResponse is the response message to convey the list of jobs.
+message ListJobsResponse {
+    // List of jobs that match the request criteria.
+    repeated Job jobs = 1;
+}
+```
+
+###### Daprd Sidecar to Scheduler Service APIs
+
+For the daprd sidecar to Scheduler Service communication, 
+![Scheduler APIs (SideCar Facing)](./resources/0004-BIRS-distributed-scheduler/sidecarToSchedulerComm.png)
+
+We will use the same exact protos from the Public Dapr API, but inside a **new**: `dapr/proto/scheduler/scheduler.proto`. However, the one difference will be in the `Job message`, where we will add:
+
+```proto
+    // Namespace for jobs to allow for multi tenancy.
+    string namespace = 5;
+
+    // The metadata associated with the job.
+    // The sidecar will create the unique `key` for storing data in the state store and pass the generated `key` along to the scheduler service for data lookup upon ‘trigger’ time later on. 
+    // The sidecar will also add metadata in order to know whether this job is registered for an actor. This is needed, as the routing mechanism for actors is different for the callback.
+    map<string,string> metadata = 6;
+```
+
+
+To allow for the daprd sidecar to register itself with the Scheduler Service upon startup, there will be a new `register` API in `dapr/proto/scheduler/scheduler.proto` with the following:
+
+```proto
+// This will allow for communication between the daprd sidecar and the Scheduler Service. (daprd sidecar -> Scheduler Service)
+service Scheduler {
+
+    ...<same user-facing protos defined above>...
+
+    // Allow for the daprd sidecar to register with a control plane Scheduler Service
+    rpc Register(RegisterRequest) returns (google.protobuf.Empty) {}
+}
+
+// RegisterRequest is the message to register a daprd sidecar with a control plane service. 
+message RegisterRequest {
+    // The id of the application (app_id). This is used as the key in etcd for the Scheduler Service.
+    string app_id = 1;
+
+    // The hostname of the daprd sidecar.
+    string hostname = 2;
+
+    // The port of the daprd sidecar.
+    int32 port = 3;
+}
+```
+
+Alternatively, there can be a streaming connection between the daprd sidecar and the Scheduler Service.
+
+###### Scheduler Service APIs
+
+In a **new** `dapr/proto/scheduler/scheduler_callback.proto`, the Scheduler Service will have a 'callback' to the app in which to fulfill the job trigger execution orchestration (ie. send back the data to the app at the time of its trigger). 
+
+```proto
+// This is intended for the Scheduler Service to call back to the daprd sidecar at the time of the job schedule... ie it's trigger time.
+service SchedulerCallback {
+
+    // Callback RPC for job schedule being at 'trigger' time
+    rpc TriggerJob(TriggerRequest) returns (google.protobuf.Empty) {}
```

Comment:
One thing you may want to consider is some sort of rate-limiting here. For example allowing a maximum number of active jobs.

Currently we don't rate-limit reminders explicitly, but given the perf bottlenecks we have in the current implementation (for example the I/O cost of executing a reminder), there's a "natural" limit there. This new design should have a much lower cost and at this stage the risk of DoS-ing an app may be real.